File: ios/runanywhere-sdks-latest/sdk/runanywhere-kotlin/modules/runanywhere-core-qhexrt/src/main/kotlin/com/runanywhere/sdk/npu/qhexrt/QHexRT.kt
package com.runanywhere.sdk.npu.qhexrt

import ai.runanywhere.proto.v1.ErrorCode
import ai.runanywhere.proto.v1.HexagonArch
import ai.runanywhere.proto.v1.ModelInfo
import ai.runanywhere.proto.v1.NpuCapability
import ai.runanywhere.proto.v1.RegisterModelFromUrlRequest
import com.runanywhere.sdk.infrastructure.logging.SDKLogger
import com.runanywhere.sdk.native.bridge.RunAnywhereBridge
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

// RunAnywhereBridge doesn't expose this one; derived the same way as its
// RAC_ERROR_* constants (signed C ABI value = negated proto-enum magnitude).
private val RAC_ERROR_BACKEND_UNAVAILABLE = -ErrorCode.ERROR_CODE_BACKEND_UNAVAILABLE.value
private val RAC_ERROR_CAPABILITY_UNSUPPORTED = -ErrorCode.ERROR_CODE_CAPABILITY_UNSUPPORTED.value

/**
 * QHexRT module — RunAnywhere's private Qualcomm Hexagon NPU backend.
 *
 * Runs prebuilt QNN context binaries on Snapdragon V75/V79/V81 NPUs, serving
 * LLM, VLM, STT and TTS through the standard SDK APIs once registered. A thin
 * wrapper over C++ backend registration; all inference lives in the C++
 * commons layer.
 *
 * ## Pre-flight
 * ```kotlin
 * val npu = QHexRT.probeNpu()
 * if (!npu.qhexrt_supported) { /* warn; fall back to CPU engines */ }
 * ```
 *
 * ## Registration
 * ```kotlin
 * QHexRT.register()   // once during bootstrap, on a supported device
 * ```
 */
object QHexRT {
    private val logger = SDKLogger("QHexRT")

    /** Current version of the QHexRT module, as reported by the native bridge. */
    val version: String
        get() {
            RunAnywhereBridge.ensureNativeLibraryLoaded()
            return if (QHexRTBridge.ensureNativeLibraryLoaded()) {
                QHexRTBridge.nativeGetVersion()
            } else {
                "unknown"
            }
        }

    /** Human-readable module name. */
    const val moduleName: String = "QHexRT"

    @Volatile
    private var isRegistered = false
    private val registrationMutex = Mutex()
    private val registrationLock = Any()

    /**
     * Probe the device's Hexagon NPU without loading QNN. Safe to call on any
     * device; returns the all-default [NpuCapability] (unknown arch,
     * `qhexrt_supported = false`) on unsupported/unknown parts or probe
     * failure.
     */
    fun probeNpu(): NpuCapability {
        RunAnywhereBridge.ensureNativeLibraryLoaded()
        if (!QHexRTBridge.ensureNativeLibraryLoaded()) {
            logger.info("QHexRT native library unavailable; reporting unsupported NPU")
            return NpuCapability()
        }
        return try {
            NpuCapability.ADAPTER.decode(QHexRTBridge.nativeProbeNpuProto())
        } catch (e: Exception) {
            logger.error("Failed to decode NPU probe proto: ${e.message}", throwable = e)
            NpuCapability()
        }
    }

    /**
     * Return whether [arch] is supported by this QHexRT build.
     *
     * The support boundary is owned by `librac_backend_qhexrt.so`; Kotlin
     * forwards the generated protobuf enum value without duplicating the
     * V75/V79/V81 set.
     */
    fun isArchitectureSupported(arch: HexagonArch): Boolean {
        RunAnywhereBridge.ensureNativeLibraryLoaded()
        if (!QHexRTBridge.ensureNativeLibraryLoaded()) return false
        return QHexRTBridge.nativeArchIsSupported(arch.value)
    }

    /**
     * Match native product catalog policy for [modelId] against [arch].
     */
    fun modelSupportsArchitecture(
        modelId: String,
        arch: HexagonArch,
    ): Boolean {
        RunAnywhereBridge.ensureNativeLibraryLoaded()
        if (!QHexRTBridge.ensureNativeLibraryLoaded()) return false
        return QHexRTBridge.nativeCatalogModelSupportsArch(modelId, arch.value)
    }

    /** Whether native product policy marks [modelId] as HF-authenticated. */
    fun modelRequiresHfAuth(modelId: String): Boolean {
        RunAnywhereBridge.ensureNativeLibraryLoaded()
        if (!QHexRTBridge.ensureNativeLibraryLoaded()) return false
        return QHexRTBridge.nativeCatalogModelRequiresHfAuth(modelId)
    }

    /**
     * Register [request] only when native product policy allows it here.
     *
     * The app remains the source of the model URL and presentation metadata.
     * QHexRT owns device probing and architecture selection, then composes the
     * shared C++ URL-registration/download pipeline. Returns `null` for the
     * normal "model is not eligible on this device" outcome.
     */
    suspend fun registerModelForDevice(request: RegisterModelFromUrlRequest): ModelInfo? {
        RunAnywhereBridge.ensureNativeLibraryLoaded()
        if (!QHexRTBridge.ensureNativeLibraryLoaded()) return null
        val bytes =
            QHexRTBridge.nativeCatalogRegisterModelProto(QHexRTCatalogWire.encodeRequest(request))
                ?: return null
        return QHexRTCatalogWire.decodeModel(bytes)
    }

    /**
     * Register the QHexRT backend with the C++ plugin registry. Suspend so
     * callers can await module bootstrap from a coroutine scope. Safe to call on
     * unsupported devices — registration is rejected and the app falls back to
     * CPU engines.
     */
    suspend fun register() {
        registrationMutex.withLock { registerInternal() }
    }

    /** Unregister the QHexRT backend from the C++ registry. */
    suspend fun unregister() {
        registrationMutex.withLock {
            if (!isRegistered) return
            unregisterNative()
            isRegistered = false
            logger.info("QHexRT backend unregistered")
        }
    }

    private fun registerInternal() {
        if (isRegistered) {
            logger.debug("QHexRT already registered, returning")
            return
        }
        logger.info("Registering QHexRT backend with C++ registry...")
        when (val result = registerNative()) {
            RunAnywhereBridge.RAC_SUCCESS,
            RunAnywhereBridge.RAC_ERROR_MODULE_ALREADY_REGISTERED,
            -> {
                isRegistered = true
                logger.info("QHexRT backend registered successfully (LLM/VLM/STT/TTS)")
            }
            RAC_ERROR_BACKEND_UNAVAILABLE,
            RAC_ERROR_CAPABILITY_UNSUPPORTED,
            -> {
                logger.info("QHexRT not registered: no supported Hexagon NPU on this device")
            }
            else -> {
                logger.error("QHexRT registration failed with code: $result")
            }
        }
    }

    /**
     * Enable auto-registration. Access this property to trigger C++ backend
     * registration once.
     */
    val autoRegister: Unit by lazy {
        synchronized(registrationLock) { registerInternal() }
    }
}

/** Byte/enum transport only; all QHexRT catalog policy stays native. */
internal object QHexRTCatalogWire {
    fun encodeRequest(request: RegisterModelFromUrlRequest): ByteArray =
        RegisterModelFromUrlRequest.ADAPTER.encode(request)

    fun decodeModel(bytes: ByteArray): ModelInfo = ModelInfo.ADAPTER.decode(bytes)
}

private val logger = SDKLogger("QHexRT")

internal fun QHexRT.registerNative(): Int {
    RunAnywhereBridge.ensureNativeLibraryLoaded()
    val skelDirectory = QHexRTSkelInstaller.installIfAvailable()
    if (!QHexRTBridge.ensureNativeLibraryLoaded()) {
        logger.info("QHexRT native library unavailable; skipping backend registration")
        return RAC_ERROR_BACKEND_UNAVAILABLE
    }
    QHexRTBridge.nativeSetSkelDirectory(skelDirectory)
    return QHexRTBridge.nativeRegister()
}

internal fun QHexRT.unregisterNative(): Int =
    if (QHexRTBridge.ensureNativeLibraryLoaded()) {
        QHexRTBridge.nativeUnregister()
    } else {
        RAC_ERROR_BACKEND_UNAVAILABLE
    }

File: ios/runanywhere-sdks-latest/sdk/runanywhere-react-native/packages/qhexrt/src/QHexRT.ts
/**
 * @runanywhere/qhexrt - QHexRT Module
 *
 * QHexRT (Qualcomm Hexagon NPU) module wrapper for the RunAnywhere React
 * Native SDK. Provides backend registration and a pre-flight NPU capability
 * probe so apps can warn unsupported devices before loading a model.
 *
 * App-owned model URLs and metadata cross the QHexRT catalog facade, which
 * selects the matching device architecture before composing core lifecycle
 * registration and download services.
 *
 * QHexRT is Qualcomm-only (Snapdragon Hexagon NPU): Android arm64 exclusively.
 * On non-Snapdragon or non-Android devices the probe reports an unsupported,
 * unknown part and register() returns false.
 */

import { QHexRTProvider } from './QHexRTProvider';
import { SDKLogger } from '@runanywhere/core/internal';
import {
  NpuCapability,
  HexagonArch,
} from '@runanywhere/proto-ts/hardware_profile';
import {
  type ModelInfo,
  type RegisterModelFromUrlRequest,
} from '@runanywhere/proto-ts/model_types';
import { QHexRTCatalogWire } from './QHexRTCatalogWire';

const log = new SDKLogger('NPU.QHexRT');

// Re-export the generated wire types so consumers never hand-mirror them.
export { NpuCapability, HexagonArch };

/**
 * The unknown/unsupported fallback used when the native probe is unavailable
 * (non-Android platforms, non-Snapdragon devices, or an older commons without
 * rac_qhexrt_probe_proto).
 */
function unknownNpuCapability(): NpuCapability {
  return NpuCapability.fromPartial({ socId: -1, archName: 'unknown' });
}

function decodeNpuCapability(buffer: ArrayBuffer | null): NpuCapability {
  if (!buffer || buffer.byteLength === 0) {
    return unknownNpuCapability();
  }
  try {
    return NpuCapability.decode(new Uint8Array(buffer));
  } catch (error) {
    log.warning(
      `Failed to decode NPU probe result: ${error instanceof Error ? error.message : String(error)}`
    );
    return unknownNpuCapability();
  }
}

/**
 * QHexRT Module
 *
 * Provides backend registration, the NPU capability probe, and device-aware
 * registration for app-owned QHexRT catalog definitions.
 *
 * ## Usage
 *
 * ```typescript
 * import { QHexRT } from '@runanywhere/qhexrt';
 * import { RunAnywhere } from '@runanywhere/core';
 *
 * const npu = await QHexRT.probeNpu();
 * if (!npu.qhexrtSupported) {
 *   // warn: this device is not in the validated V75/V79/V81 set
 * }
 * await QHexRT.register();
 * ```
 */
export const QHexRT = {
  HexagonArch,

  /**
   * Register the QHexRT module with the SDK.
   * Registers the LLM, VLM, STT, and TTS providers with the C++ registry.
   */
  async register(): Promise<boolean> {
    log.debug('Registering QHexRT module');
    const registered = await QHexRTProvider.register();
    if (registered) {
      log.info('QHexRT module registered');
    }
    return registered;
  },

  /**
   * Unregister the QHexRT module from the SDK.
   */
  async unregister(): Promise<boolean> {
    log.info('Unregistering QHexRT module');
    return QHexRTProvider.unregister();
  },

  /**
   * Check if this module is registered with the native backend registry.
   */
  async isRegistered(): Promise<boolean> {
    return QHexRTProvider.isRegistered();
  },

  /**
   * Pre-flight probe of the device's Qualcomm Hexagon NPU capability.
   * Does NOT load QNN or the engine. Decodes the serialized
   * `runanywhere.v1.NpuCapability` proto emitted by QHexRT's
   * rac_qhexrt_probe_proto(). Returns the unknown/unsupported fallback
   * (socId -1, archName "unknown") when the native module is unavailable
   * (e.g. non-Snapdragon devices).
   */
  async probeNpu(): Promise<NpuCapability> {
    const raw = await QHexRTProvider.probeNpuRaw();
    return decodeNpuCapability(raw);
  },

  /** The native QHexRT support policy; no V75/V79/V81 set is copied in TS. */
  isArchitectureSupported(arch: HexagonArch): boolean {
    return QHexRTProvider.isArchitectureSupported(arch);
  },

  /** Match QHexRT native product policy for a model against an architecture. */
  modelSupportsArchitecture(modelId: string, arch: HexagonArch): boolean {
    return QHexRTProvider.modelSupportsArchitecture(modelId, arch);
  },

  /** Whether QHexRT native product policy marks a model HF-authenticated. */
  modelRequiresHfAuth(modelId: string): boolean {
    return QHexRTProvider.modelRequiresHfAuth(modelId);
  },

  /**
   * Register an app-owned URL/model definition only when it matches the
   * current device. Native QHexRT owns probing and chip selection, then
   * composes commons' shared registration/download pipeline. `null` is the
   * normal ineligible model/device outcome.
   */
  async registerModelForDevice(
    request: RegisterModelFromUrlRequest
  ): Promise<ModelInfo | null> {
    const raw = await QHexRTProvider.registerModelForDeviceRaw(
      QHexRTCatalogWire.encodeRequest(request)
    );
    return raw ? QHexRTCatalogWire.decodeModel(raw) : null;
  },
};
